pkg/proto/impl.go

```diff
@@ -164,46 +167,71 @@ func (p *protocol) handleMessage(peerMsg p2p.PeerMessage) error {
 	}
 	if msg.HashList != nil {
 		log.Log().Debugf("Received hash list from peer (peer=%s)", peer)
-		for _, hashInTime := range msg.HashList.Hashes {
-			hash := model.Hash(hashInTime.Hash)
-			p.hashSource.AddDocumentHash(hash, time.Unix(0, hashInTime.Time))
+		for _, current := range msg.HashList.Hashes {
+			hash := model.Hash(current.Hash)
+			if hash.Empty() {
+				log.Log().Warn("Received document doesn't contain a hash, skipping.")
+			}
+			if current.Time == 0 {
+				log.Log().Warnf("Received document doesn't contain a timestamp, skipping (hash=%s).", hash)
+			}
 			if !p.hashSource.HasDocument(hash) {
-				log.Log().Infof("Received document hash from peer that we don't have yet, will query it (peer=%s,hash=%s)", peer, model.Hash(hash))
+				document := model.Document{
+					Type:      current.Type,
+					Timestamp: time.Unix(0, current.Time),
+					Hash:      current.Hash,
+				}
+				p.hashSource.AddDocument(document)
+			}
+			if !p.hashSource.HasContentsForDocument(hash) {
```

Comment:
You could also put in on a channel and be done with it here!

Reply:
For what purpose, abstraction/concurrency/..?

the call to `p2pNetwork.Send(peer, message)` puts the messages on a channel to be sent to that peer, so this routine won't be blocked.

Reply:
> You could also put in on a channel and be done with it here!

For what purpose, abstraction/concurrency/..?

the call to `p2pNetwork.Send(peer, message)` puts the messages on a channel to be sent to that peer, so this routine won't be blocked.